### DISCUSSION TITLE
How to render MDX file in TAB addon?

### DISCUSSION THREAD

**Discussion by TheColorRed** [storybookjs/storybook#18211] [Help]
I am trying create an addon that adds a tab and should display an MDX file. The problem that I am having is that I am getting an error:

> Uncaught Error: MDXContent(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null.

```javascript
import { addons, types } from '@storybook/addons';
import React from 'react';


addons.register('my-api', (api) => {
  addons.add('my-api/tab', {
    type: types.TAB,
    title: 'API',
    route: ({ storyId, refId }) => (refId ? `/api/${refId}_${storyId}` : `/api/${storyId}`),
    match: ({ viewMode }) => viewMode === 'api',
    render: () => {
      const Page = api.getCurrentStoryData()?.parameters?.api?.page;
      // console.table('md', typeof Page, Page, typeof Page === 'function' ? Page() : undefined);
      return (
        <div>
          {
            typeof Page === 'undefined' ?
              <h2>No API setup for this Story.</h2> :
              <Page />
          }
        </div>
      );
    },
  });
});
```

Note: I have also tried this: `<MDXProvider><Page /></MDXProvider>`

Then in the story I have this:

```typescript
import Api from './button.api.mdx';

export default {
  parameters: {
    api: { page: Api }
  }
} as Meta;
```

What is the proper way to handling rendering of MDX files in a new tab?

**Comment by TheColorRed** [comment]
Is this a bug or maybe I am using an outdated version?

**Comment by AntonisFK** [comment]
@TheColorRed, I'm running into the same issue. Did you ever solve this?

**Reply by TheColorRed** [reply]
No I have not found how to do it.

**Comment by kylegach** [comment]
I believe you'll need to compile the MDX yourself. This all runs in the browser, so [`evaluate`](https://mdxjs.com/packages/mdx/#evaluatefile-options) (as [recommended by their docs](https://mdxjs.com/docs/getting-started/#bundler)) is your only option.

**Reply by AntonisFK** [reply]
I'm able to use evaluate a function on MD files but when I pass an MDX file like above it fails to work. When I console log the content in my addon it doesn't look like an MDX function it returns

```js
ƒ result() {
        // eslint-disable-next-line no-eval
        var f = eval("(".concat(sourceSanitized, ")"));
        return f.apply(void 0, arguments);
      }

```

I'm assuming it has to do something with the way the manager (Storybook UI code) and preview (user code) communicate with each other.  Explained in this [comment](https://github.com/storybookjs/storybook/discussions/13922#discussioncomment-372348)
